modules/service/trail-service.js

```diff
@@ -0,0 +1,106 @@
+const Utilities = require('../Utilities');
+const constants = require('../constants');
+
+class TrailService {
+    constructor(ctx) {
+        this.graphStorage = ctx.graphStorage;
+        this.importService = ctx.importService;
+        this.log = ctx.logger;
+    }
+
+    async lookupTrail(identifierTypes, identifierValues, opcode) {
+        if (Utilities.arrayze(identifierTypes).length !==
+            Utilities.arrayze(identifierValues).length) {
+            throw Error('Identifier array length mismatch');
+        }
+
+        if (!['EQ', 'IN'].includes(opcode)) {
+            throw Error('Invalid opcode parameter. Opcode can be EQ or IN');
+        }
+
+        const identifierKeys = [];
+
+        const typesArray = Utilities.arrayze(identifierTypes);
+        const valuesArray = Utilities.arrayze(identifierValues);
+
+        for (let i = 0; i < typesArray.length; i += 1) {
+            identifierKeys.push(Utilities.keyFrom(typesArray[i], valuesArray[i]));
+        }
+
+        const result = await this.graphStorage.lookupTrail({
+            identifierKeys,
+            opcode,
+        });
+
+        return result;
+    }
+
+    async findTrail(
+        uniqueIdentifiers, depth, reach,
+        includedConnectionTypes,
+        excludedConnectionTypes,
+    ) {
+        if (includedConnectionTypes && excludedConnectionTypes && includedConnectionTypes.find(x => excludedConnectionTypes.includes(x))) {
+            throw Error('Included and excluded connection types contain same types');
+        }
+
+        const trail =
+            await this.graphStorage.findTrail({
+                uniqueIdentifiers,
+                depth,
+                includedConnectionTypes,
+                excludedConnectionTypes,
+            });
+
+        let response = this.importService.packTrailData(trail);
+
+        if (reach === constants.TRAIL_REACH_PARAMETERS.extended) {
+            response = await this._extendResponse(response);
+        }
+
+        return response;
+    }
+
+    async _extendResponse(response) {
+        const missingObjects = {};
+        for (const trailElement of response) {
+            const object = trailElement.otObject;
+
+            const elementIsMissing =
+                (array, element) => !array.find(e => e.otObject['@id'] === element['@id']);
+
+            for (const relation of object.relations) {
+                if (elementIsMissing(response, relation.linkedObject)) {
+                    if (!missingObjects[relation.linkedObject['@id']]) {
+                        missingObjects[relation.linkedObject['@id']] = trailElement.datasets;
+                    } else {
+                        missingObjects[relation.linkedObject['@id']] =
```

Comment:
I know this is not the point of the PR but we could really optimize here if we were to change the top condition to 
`if (missingObjects[relation.linkedObject['@id']] || elementIsMissing...` because that way we don't need to iterate over the entire array if we already know the object is missing